oauth2/src/main/java/com/linecorp/armeria/server/auth/oauth2/OAuth2AuthorizationFailureHandler.java

```diff
@@ -0,0 +1,188 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.server.HttpService;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.auth.AuthFailureHandler;
+import com.linecorp.armeria.server.auth.AuthServiceBuilder;
+import com.linecorp.armeria.server.auth.Authorizer;
+
+/**
+ * A callback which is invoked to handle OAuth 2.0 authorization failure indicated by {@link Authorizer}.
+ * Composes OAuth 2.0 authorization error response in one of the following ways:
+ * <ul>
+ *    <li>
+ *    invalid_request
+ *          <p>The request is missing a required parameter, includes an
+ *          unsupported parameter or parameter value, repeats the same
+ *          parameter, uses more than one method for including an access
+ *          token, or is otherwise malformed. The resource server SHOULD
+ *          respond with the HTTP 400 (Bad Request) status code.</p>
+ *          Example:
+ *          <pre>{@code
+ *              HTTP/1.1 400 Bad Request
+ *              Content-Type: application/json;charset=UTF-8
+ *              {"error":"unsupported_token_type"}
+ *          }</pre>
+ *    </li>
+ *    <li>
+ *    invalid_token
+ *          <p>The access token provided is expired, revoked, malformed, or
+ *          invalid for other reasons. The resource SHOULD respond with
+ *          the HTTP 401 (Unauthorized) status code. The client MAY
+ *          request a new access token and retry the protected resource
+ *          request.</p>
+ *          Example:
+ *          <pre>{@code
+ *              HTTP/1.1 401 Unauthorized
+ *              WWW-Authenticate: Bearer realm="example",
+ *                                error="invalid_token",
+ *                                scope="read write"
+ *          }</pre>
+ *    </li>
+ *    <li>
+ *    insufficient_scope
+ *          <p>The request requires higher privileges than provided by the
+ *          access token. The resource server SHOULD respond with the HTTP
+ *          403 (Forbidden) status code and MAY include the "scope"
+ *          attribute with the scope necessary to access the protected
+ *          resource.</p>
+ *          Example:
+ *          <pre>{@code
+ *              HTTP/1.1 403 Forbidden
+ *              Content-Type: application/json;charset=UTF-8
+ *              {"error":"insufficient_scope"}
+ *          }</pre>
+ *    </li>
+ * </ul>
+ *
+ * @see AuthServiceBuilder#onFailure(AuthFailureHandler)
+ */
+public class OAuth2AuthorizationFailureHandler implements AuthFailureHandler {
```

Comment:
Does it make sense to hide this from the public API?

Reply:
fixed